mlir/lib/Catalyst/Transforms/BufferizableOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,236 @@
+#include "mlir/Dialect/Bufferization/IR/BufferizableOpInterface.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Bufferization/Transforms/Bufferize.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "Catalyst/IR/CatalystOps.h"
+#include "Catalyst/Transforms/BufferizableOpInterfaceImpl.h"
+
+using namespace mlir;
+using namespace catalyst;
+
+namespace {
+
+/// Bufferization of catalyst.print. Get memref of printOp.val.
+struct PrintOpInterface
+    : public bufferization::BufferizableOpInterface::ExternalModel<PrintOpInterface,
+                                                    PrintOp> {
+    bool bufferizesToMemoryRead(Operation *op, OpOperand &opOperand,
+                                const bufferization::AnalysisState &state) const {
+        return true;
+    }
+
+    bool bufferizesToMemoryWrite(Operation *op, OpOperand &opOperand,
+                                 const bufferization::AnalysisState &state) const {
+        return false;
+    }
+
+    bufferization::AliasingValueList getAliasingValues(Operation *op,
+                                        OpOperand &opOperand,
+                                        const bufferization::AnalysisState &state) const {
+        return {};
+    }
+
+    LogicalResult bufferize(Operation *op, RewriterBase &rewriter,
+                            const bufferization::BufferizationOptions &options) const {
+        auto printOp = cast<PrintOp>(op);
+        if (printOp.getVal()) {
+            FailureOr<Value> source = getBuffer(rewriter, printOp.getVal(), options);
+            if (failed(source))
+                return failure();
+            bufferization::replaceOpWithNewBufferizedOp<PrintOp>(rewriter, op, *source,
+                            printOp.getConstValAttr(), printOp.getPrintDescriptorAttr());
+        }
+        return success();
+    }
+};
+
+/// Bufferization of catalyst.print. Mainly get buffers for arguments.
+struct CustomCallOpInterface
+    : public bufferization::BufferizableOpInterface::ExternalModel<CustomCallOpInterface,
+                                                    CustomCallOp> {
+    bool bufferizesToMemoryRead(Operation *op, OpOperand &opOperand,
+                                const bufferization::AnalysisState &state) const {
+        return true;
+    }
+
+    bool bufferizesToMemoryWrite(Operation *op, OpOperand &opOperand,
+                                 const bufferization::AnalysisState &state) const {
+        return false;
```

Comment:
Can you change this to true and give it a try.

Reply:
I found this, [here](https://m-sp.org/downloads/llvm_dev_2023.pdf)

```
Common Pitfalls
- BufferizableOpInterface external models not
registered: ops are not getting bufferized
- Function boundary bufferization not enabled:
function bbArgs are read-only
- getBufferType() not implemented: mismatch
between loop inits and iter_arg types
- getBufferType() / bufferize() ignores the
memory space of MemRef types
- Input IR has to_tensor without restrict
- Assumptions that a computation materializes in a
certain buffer without making it explicit (e.g.,
tensor ops may canonicalize/fold away)
- Loop op yields value that is not equivalent to
corresponding iter_arg: inefficient due to current
implementation details
```

Reply:
https://discord.com/channels/636084430946959380/642426447167881246/1280550112002572380

Reply:
It might be useful to include the relevant snippet rather than just the discord link :)

Reply:
> Function boundary bufferization not enabled:
function bbArgs are read-only

I sometimes like to do a quick check if it takes less than a minute to try it out. This could be twisted a little bit as having FunctionOpInterface's have write arguments, but it was not very likely. 

> It might be useful to include the relevant snippet rather than just the discord link :)

The real reason was that bufferization does not work yet on FunctionOpInterfaces, only on FuncOp's. The maintainers have said that changing the concrete ops for interfaces is something that can be upstreamed and Tzung-Han is currently working on it.

Reply:
> I sometimes like to do a quick check if it takes less than a minute to try it out. This could be twisted a little bit as having FunctionOpInterface's have write arguments, but it was not very likely.

Got it, was just trying to understand so I can learn more about bufferization :) (and definitely nothing wrong with trying things out)

> The real reason was that bufferization does not work yet on FunctionOpInterfaces, only on FuncOp's. The maintainers have said that changing the concrete ops for interfaces is something that can be upstreamed and Tzung-Han is currently working on it.

Glad we figured it out :D